### ISSUE TITLE
Issues with virtual machines/containers and network

### ISSUE THREAD

**Issue by Fornball** [cb-linux/breath#235]
Apologies for being very vague, the problem is I couldnt really get to the root of the problem. I am not sure if its just my system, but could you/someone else try running any virtual machine with qemu and see if network works?

I am having issues with virtual machine/containers and network. Wifi/Ethernet works fine on the system itself, but when I try creating a vm or a container, they have issues with network.

- Waydroid via an lxc container:
I have tried installing waydroid, but I keep getting an issue with networking. The issue is described [here](https://github.com/waydroid/waydroid/issues/346), its where lxc tries to create a container but failes because it cant create a network for it(or something like that). I have tried all the possible solutions, but it didnt work.
- Qemu/kvm with virt manager:
When creating a virtual machine with network enabled(which it is by default), it fails by saying I'm missing a package. But every package I need is installed and it works fine on my desktop pc(Pop!_OS 22.04) with the same packages installed. The error I'm getting is: ```libvirt.libvirtError: Unable to create bridge virbr0: Package not installed```

One thing that both of these problems have in common, is that some solutions talk about some kernel module or a kernel parameter(I'm not sure which one tho).

**Comment by lauretandretheo** [comment]
`libvirt.libvirtError: Unable to create bridge virbr0: Package not installed` is due to a lack of kernel support.

**Comment by Fornball** [comment]
So there is not fix for it?

**Comment by runcros** [comment]
> So there is not fix for it?
 The maintainer will answer you, I think it's possible with building a kernel without lack of support but I don't know how.

**Comment by MilkyDeveloper** [comment]
This is fixable by changing the kernel config. I'll do it now.

**Comment by runcros** [comment]
Just to check, is the package libvirt0 is installed ?

**Comment by Fornball** [comment]
Yes, all these packages are installed on the chromebook

**Comment by runcros** [comment]
Please also check for all the packages listed here : https://wiki.archlinux.org/title/Libvirt#Server

**Comment by runcros** [comment]
If all package are installed, try also to install this package `qemu-block-gluster` then check if issue appear again.

**Comment by Fornball** [comment]
Ehh, how do I install it?
```E: Unable to locate package qemu-block-gluster```

**Comment by runcros** [comment]
Try `qemu-block-extra`

**Comment by runcros** [comment]
Also send the output of `getent group kvm`

**Comment by Fornball** [comment]
qemu-block-extra was already installed.
```
$ getent group kvm
kvm:x:109:
```

**Comment by runcros** [comment]
You should try this fix https://bbs.archlinux.org/viewtopic.php?pid=1728381#p1728381

If it doesn't work, wait for my pr to be pulled and then update the kernel after the release.

**Comment by Fornball** [comment]
The pr just got merged, how do I update the kernel on a non usb install?

**Comment by runcros** [comment]
The github action rebuild has finished, you can update the kernel.

**Comment by Fornball** [comment]
Updated kernel on internal install, rebooted, tried waydroid and a virtual machine, didnt work. Created a fresh live usb, plugged it in, installed waydroid first, same error. Tried a virtual machine, same thing as before.

**Comment by runcros** [comment]
Can you replace your /usr/local/bin/updatekernel-on-emmc with the following file (remove .txt) 
[updatekernel-on-emmc.txt](https://github.com/cb-linux/breath/files/8984512/updatekernel-on-emmc.txt)
It will update your kernel from my repository (I've added others kernel module)

**Comment by Fornball** [comment]
Updated, rebooted, same errors

**Comment by runcros** [comment]
Okay I have added others kernel module (again !) try to update with the previous file. 
If the problem is still not solved. We should look elsewhere than in the kernel modules

**Comment by Fornball** [comment]
No, unfortunately still the same errors

**Comment by runcros** [comment]
I added all kernel modules needed by KVM so I think the problem is elsewhere.

**Comment by dinoswong** [comment]
I've updated the latest version of kernel. And I tried to start a VM with NAT network by using command qemu. The guest also cannot access the external network. And I found it missing kernel module ipt_MASQUERADE. Would you include it by default? Thanks.

**Comment by runcros** [comment]
I've added these module (you can see in the commit : https://github.com/runcros/breath/commit/ada1a1c4b13b5d4b49886cd431dd7aaaef8336fc)

CI will take ~2 hours you can try if it works after the release on my repository.
Here is my updatekernel-on-emmc to update from my repo https://github.com/cb-linux/breath/files/8984512/updatekernel-on-emmc.txt

**Comment by runcros** [comment]
You can update.

**Comment by dinoswong** [comment]
Thanks. I've updated kernel. It was built at 26-Jun **13:06:05. But it still missing the modules. 

The modules should be located on lib/modules/5.10.106/kernel/net/ipv4/netfilters. But there is no such that directory and file. I've download modules.tar.xz manually and cannot find it too.

**Comment by runcros** [comment]
Can you give me the exact name of the module it should looks something like this :
```
CONFIG_AUDIT=y
CONFIG_HAVE_ARCH_AUDITSYSCALL=y
CONFIG_AUDITSYSCALL=y
```
I don't know what is the time lol I am on GMT+4 the kernel was built 3 hours ago.

**Comment by dinoswong** [comment]
Sorry, it is my fault. The module has been changed to xt_MASQUERADE. It is included in the new build. My VM can access the external network via NAT. Thanks for help.

**Comment by runcros** [comment]
@Fornball is it working for you ?

**Comment by Fornball** [comment]
No, but the error is different now. For some reason breath is using nftables-legacy instead of the normal ones.
```
libvirt.libvirtError: internal error: Failed to apply firewall rules /usr/sbin/iptables -w --table filter --list-rules: # Warning: iptables-legacy tables present, use iptables-legacy to see them
iptables v1.8.7 (nf_tables): Could not fetch rule set generation id: Invalid argument
```

**Comment by dinoswong** [comment]
> 

I've changed the iptables to use the /usr/sbin/iptables-legacy. Under ChromeOS, it require to change it too.

**Comment by runcros** [comment]
@Fornball if the fix works for you too I will add it to docs

**Comment by Fornball** [comment]
> I've changed the iptables to use the /usr/sbin/iptables-legacy instead of /usr/sbin/iptables-nft . Under ChromeOS, it require to change it too.

How do I do that?

**Comment by dinoswong** [comment]
> > I've changed the iptables to use the /usr/sbin/iptables-legacy instead of /usr/sbin/iptables-nft . Under ChromeOS, it require to change it too.
> 
> How do I do that?

 update-alternatives --config iptables

**Comment by Fornball** [comment]
virt manager now gives me this error: ```libvirt.libvirtError: Unable to open /dev/net/tun, is tun module loaded?: No such device
``` which is supposed to resolve itself by rebooting, but I have rebooted twice now and its still there.
I changed the iptables to /usr/sbin/iptables-legacy with the ```sudo update-alternatives --config iptables``` command but waydroid still shows the same errors.

**Comment by runcros** [comment]
> virt manager now gives me this error: ```libvirt.libvirtError: Unable to open /dev/net/tun, is tun module loaded?: No such device``` which is supposed to resolve itself by rebooting, but I have rebooted twice now and its still there.
> I changed the iptables to /usr/sbin/iptables-legacy with the ```sudo update-alternatives --config iptables``` command but waydroid still shows the same errors with this ```iptables v1.8.7 (legacy): unknown option "--checksum-fill"
> Try `iptables -h' or 'iptables --help' for more information.```

Have you tried on clean install ?

**Comment by dinoswong** [comment]
> virt manager now gives me this error: `libvirt.libvirtError: Unable to open /dev/net/tun, is tun module loaded?: No such device` which is supposed to resolve itself by rebooting, but I have rebooted twice now and its still there. I changed the iptables to /usr/sbin/iptables-legacy with the `sudo update-alternatives --config iptables` command but waydroid still shows the same errors with this `` iptables v1.8.7 (legacy): unknown option "--checksum-fill" Try `iptables -h' or 'iptables --help' for more information. ``

It seem you haven't load tun module. Please execute the command 'sudo modprobe tun'.

**Comment by runcros** [comment]
Should work with Breath on UEFI can you test it ? Please join the discord for information about https://discord.gg/WatsaYhXCa

**Comment by apacelus** [comment]
Lxc containers still wont start(preventing me from running waydroid)
